src/Marten/Events/Daemon/Coordination/ProjectionCoordinator.cs

```diff
@@ -0,0 +1,252 @@
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using JasperFx.Core;
+using Marten.Events.Daemon.Resiliency;
+using Marten.Storage;
+using Microsoft.Extensions.Logging;
+using Polly;
+
+namespace Marten.Events.Daemon.Coordination;
+
+public class ProjectionCoordinator<T>: ProjectionCoordinator, IProjectionCoordinator<T> where T : IDocumentStore
+{
+    public ProjectionCoordinator(T documentStore, ILogger<ProjectionCoordinator> logger) : base(documentStore, logger)
+    {
+    }
+}
+
+public class ProjectionCoordinator : IProjectionCoordinator
+{
+    private readonly StoreOptions _options;
+    private readonly ILogger<ProjectionCoordinator> _logger;
+
+    private readonly ConcurrentDictionary<IMartenDatabase, IProjectionDaemon> _daemons = new();
+    private readonly ResiliencePipeline _resilience;
+    private readonly CancellationTokenSource _cancellation = new();
+    private Task _runner;
+
+    public ProjectionCoordinator(IDocumentStore documentStore, ILogger<ProjectionCoordinator> logger)
+    {
+        var store = (DocumentStore)documentStore;
+
+        if (store.Options.Projections.AsyncMode == DaemonMode.Solo)
+        {
+            Distributor = new SoloProjectionDistributor(store);
+        }
+        else if (store.Options.Projections.AsyncMode == DaemonMode.HotCold)
+        {
+            if (store.Options.Tenancy is DefaultTenancy)
+            {
+                Distributor = new SingleTenantProjectionDistributor(store);
+            }
+            else
+            {
+                Distributor = new MultiTenantedProjectionDistributor(store);
+            }
+        }
+
+        _options = store.Options;
+        _logger = logger;
+        _resilience = store.Options.ResiliencePipeline;
+
+        Store = store;
+    }
+
+    public IProjectionDaemon DaemonForMainDatabase()
+    {
+        var database = (MartenDatabase)Store.Tenancy.Default.Database;
+        if (_daemons.TryGetValue(database, out var daemon))
```

Comment:
**SUGGESTION:** I think that it'd be safer to use here `GetOrAdd`. We're not doing here async operations, so it's safer to use as we won't have race conditions between TryGetValue and assignments calls.

Reply:
See also https://gist.github.com/davidfowl/3dac8f7b3d141ae87abf770d5781feed#file-concurrentdictionaryextensions-cs-L53.

Reply:
At this point I'm more comfortable with ImHashMap + a lock and I'm switching that now like you suggested before.